### PR TITLE
Format tests

### PR DESCRIPTION
**Pull request by mmatera**
This PR incorporates an extended set of tests for expression formatting. This was intended to check that #181 behaves equally or better than master. 

<a href="https://gitpod.io/#https://github.com/Mathics3/mathics-core/pull/184"></a>

### CODE REVIEW

File: test/test_format.py [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+    [
+        ('"4"', "4", None),
+        ("4", "4", None),
+        ('"4.32"', "4.32", None),
+        ('"4.32313213213`2"', "4.32313213213`2", None),
+        ('"Hola!"', "Hola!", None),
+        ("a", "a", None),
+        ("Pi", "Pi", None),
+        ("a^4", "a^4", None),
+        # Boxerror
```
Comment (mmatera) [inline test/test_format.py:262]: These tests passes without problem in #181

File: test/test_format.py [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+
+# MathML StandardForm
+@pytest.mark.parametrize(
+    ("str_expr", "str_expected", "msg"),
+    [
+        ('"4"', "<mtext>4</mtext>", None),
+        # The following seems wrong: they should be formatted as numbers
+        ("4", "<mn>4</mn>", None),
+        ('"4.32"', "<mtext>4.32</mtext>", None),
+        ('"4.32313213213`2"', "<mtext>4.32313213213`2</mtext>", None),
```
Comment (mmatera) [inline test/test_format.py:20]: This behavior is wrong: the output should be ``<mtext>4.32`2</mtext>``.

File: test/test_format.py [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+        # Boxerror
+        #        ("Subscript[a, 4]", "Subscript[a, 4]", None),
+        #        ("Subsuperscript[a, p, q]", "Subsuperscript[a, p, q]", None),
+        #        (
+        #            "Integrate[F[x],{x,a,g[b]}]",
+        #            "Subsuperscript[∫, a, g[b]]\u2062F[x]\u2062\uf74cx",
+        #            None,
+        #        ),
+        # This seems to be wrong...
+        ("a^(b/c)", "a^ ( b ) / ( c ) ", None),
```
Comment (mmatera) [inline test/test_format.py:271]: This is also wrong, and is something to be fixed.

File: test/test_format.py [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+        ("1/(1+1/(1+1/a))", " ( 1 ) / ( 1+ ( 1 ) / ( 1+ ( 1 ) / ( a )  )  ) ", None),
+        # Boxerror
+        #        (
+        #            "Sqrt[1/(1+1/(1+1/a))]",
+        #            "Sqrt[ ( 1 ) / ( 1+ ( 1 ) / ( 1+ ( 1 ) / ( a )  )  ) ]",
+        #            None,
+        #        ),
+        ("Graphics[{}]", "-Graphics-", None),
+        # These tests requires ``evaluation`` as a parameter.
```
Comment (mmatera) [inline test/test_format.py:280]: These expressions require that an `evaluation` parameter be passed to format to produce the expression. This parameter should be optional.

File: test/test_format.py (hunk 3 above)
Comment (TiagoCavalcante) [reply test/test_format.py:271]: Yeah, this seems pretty wrong

**Comment by TiagoCavalcante** [comment]
@mmatera looks great! Thanks for adding more tests!

File: test/test_format.py (hunk 3 above)
Comment (mmatera) [reply test/test_format.py:271]: This is fixed in #181

File: test/test_format.py (hunk 2 above)
Comment (rocky) [reply test/test_format.py:20]: If the `('"4.32313213213`2"', "<mtext>4.32313213213`2</mtext>", None),` is wrong we shouldn't be testing for wrong output. 

We can do something like

```
TODO_TESTS=[('"4.32313213213`2"', "<mtext>4.32`2</mtext>", None), ...] 
```

and remove the wrong test from the list.

**Comment by mmatera** [comment]
@rocky, I tried to be more verbose in the motivation of each test. In the future, when the xfail tag can be removed, I would like to put all the tests associated with the same expression in an array/dict containing the test for all the formats and reduce the number of lines of this (and comments). Now it

**Comment by TiagoCavalcante** [comment]
@mmatera if you do like the following there won't be merge commits:
```bash
git checkout master
git pull
git checkout format_tests
git rebase main
# Git will tell you if there are conflicts
git push -f
```

**Comment by mmatera** [comment]
> @mmatera if you do like the following there won't be merge commits:
> 
> ```shell
> git checkout master
> git pull
> git checkout format_tests
> git rebase main
> # Git will tell you if there are conflicts commit by commit, you fix conflicts and it is done!
> git push -f
> ```

I rebased this, but the idea is to keep this close to the current main.

**Comment by TiagoCavalcante** [comment]
> I rebased this, but the idea is to keep this close to the current main.

@mmatera I'm not sure if I understood, the commands I recommended keep this close to main and don't have merge commits.

**Comment by mmatera** [comment]
@TiagoCavalcante, in this branch I just have changes in a single file, so when we want to polish it to be merged, the squash is going to be easy. The point to having all the other files close as possible to the main is to see if the tests still pass with the last changes.

**Comment by TiagoCavalcante** [comment]
> The point to having all the other files close as possible to the main is to see if the tests still pass with the last changes.

Yeah, `git rebase master` does that.

**Comment by TiagoCavalcante** [comment]
This is already pretty big, what do you think of merging it now?

**Comment by rocky** [comment]
In light of #247 is the top-level test directory the right place to put this?

File: test/test_format.py [hunk 5]
```diff
@@ -8,523 +8,789 @@
...
-    # Atoms are not still correctly processed as BoxConstruct
-    #    assert isinstance(format_result,  BoxConstruct)
-    if msg:
-        assert (
-            format_result.boxes_to_text(evaluation=session.evaluation) == str_expected
-        ), msg
-    else:
-        str_format = format_result.boxes_to_text(evaluation=session.evaluation)
-        assert str_format == str_expected
+MATHML_STRICT = True
```
Comment (rocky) [inline test/test_format.py:25]: Better would be to allow setting this from an environment variable, e.g.

```
MATHML_STRICT = int(os.environ.get("MATHML_STRICT", "1"))  # To set to false set ENV var to "0"
```

File: test/test_format.py [hunk 6]
```diff
@@ -8,523 +8,789 @@
...
+    },
+    # String with special characters
+    '"\\[Pi] is a trascendent number"': {
+        "msg": "A String",
+        "text": {
+            "System`StandardForm": "π is a trascendent number",
+            "System`TraditionalForm": "π is a trascendent number",
+            "System`InputForm": "π is a trascendent number",
+            "System`OutputForm": "π is a trascendent number",
+        },
```
Comment (rocky) [inline test/test_format.py:172]: All of the test above don't seem that fragile. I is just when we start getting to stuff like 

```
"<mtext>π&nbsp;is&nbsp;a&nbsp;trascendent&nbsp;number</mtext>",
``` 

or worse: 

````
"System`StandardForm": '<mrow><msubsup><mo>∫</mo> <mi>a</mi> <mrow><mi>g</mi> <mo>[</mo> <mi>b</mi> <mo>]</mo></mrow></msubsup> <mo form="prefix" lspace="0" rspace="0.2em">\u2062</mo> <mrow><mi>F</mi> <mo>[</mo> <mi>x</mi> <mo>]</mo></mrow> <mo form="prefix" lspace="0" rspace="0.2em">\u2062</mo> <mrow><mtext>\uf74c</mtext> <mi>x</mi></mrow></mrow>',
            "System`TraditionalForm": '<mrow><msubsup><mo>∫</mo> <mi>a</mi> <mrow><mi>g</mi> <mo>(</mo> <mi>b</mi> <mo>)</mo></mrow></msubsup> <mo form="prefix" lspace="0" rspace="0.2em">\u2062</mo> <mrow><mi>F</mi> <mo>(</mo> <mi>x</mi> <mo>)</mo></mrow> <mo form="prefix" lspace="0" rspace="0.2em">\u2062</mo> <mrow><mtext>\uf74c</mtext> <mi>x</mi></mrow></mrow>',
            "System`InputForm": "<mrow><mi>Integrate</mi> <mo>[</mo> <mrow><mrow><mi>F</mi> <mo>[</mo> <mi>x</mi> <mo>]</mo></mrow> <mtext>,&nbsp;</mtext> <mrow><mo>{</mo> <mrow><mi>x</mi> <mtext>,&nbsp;</mtext> <mi>a</mi> <mtext>,&nbsp;</mtext> <mrow><mi>g</mi> <mo>[</mo> <mi>b</mi> <mo>]</mo></mrow></mrow> <mo>}</mo></mrow></mrow> <mo>]</mo></mrow>",
            "System`OutputForm": "<mrow><mi>Integrate</mi> <mo>[</mo> <mrow><mrow><mi>F</mi> <mo>[</mo> <mi>x</mi> <mo>]</mo></mrow> <mtext>,&nbsp;</mtext> <mrow><mo>{</mo> <mrow><mi>x</mi> <mtext>,&nbsp;</mtext> <mi>a</mi> <mtext>,&nbsp;</mtext> <mrow><mi>g</mi> <mo>[</mo> <mi>b</mi> <mo>]</mo></mrow></mrow> <mo>}</mo></mrow></mrow> <mo>]</mo></mrow>
```
or even more bad:

```
System`StandardForm": '<mglyph width="294px" height="350px" src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjEuMHB4IiBoZW... 
``` 

where it is annoying. (Personally, I dislike all the extra work this is going to cause me/us when thing change, but I guess I'll just live with it.)

But  using your new-found skills of trying to find the _smallest_ example that exhibits the feature you want to test would be most appreciated here.

File: CHANGES.rst [hunk 7]
```diff
@@ -35,6 +35,7 @@ New Builtins
 * ``TraceEvaluation[]`` shows expression name calls and return values of it argument.
    -  Pass option ``ShowTimeBySteps``, to show accumulated time before each setp
```
Comment (rocky) [inline CHANGES.rst:36]: setp -> step

**Comment by mmatera** [comment]
superseded by #302

**Pull request closed without merging**